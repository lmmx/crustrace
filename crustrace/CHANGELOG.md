# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.4](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.3...crustrace-v0.1.4) - 2025-06-09

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.3](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.2...crustrace-v0.1.3) - 2025-06-08

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.2](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.1...crustrace-v0.1.2) - 2025-06-07

### <!-- 9 -->Other

- updated the following local packages: crustrace-core

## [0.1.1](https://github.com/lmmx/crustrace/compare/crustrace-v0.1.0...crustrace-v0.1.1) - 2025-06-07

### <!-- 2 -->Bug Fixes

- amend README metadata

### <!-- 9 -->Other

- license

## [0.1.0](https://github.com/lmmx/crustrace/releases/tag/crustrace-v0.1.0) - 2025-05-27

### <!-- 1 -->Features

- support impl Struct methods
- Initial commit

### <!-- 5 -->Refactor

- rename 'trace all' to 'omni', it's cleaner
- merge crustrace-attributes crate into crustrace-core ([#3](https://github.com/lmmx/crustrace/pull/3))

### <!-- 8 -->Styling

- lint

### <!-- 9 -->Other

- prepare for publishing crustrace crate
- set up git hooks
