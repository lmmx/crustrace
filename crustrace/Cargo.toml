[package]
description = "Function and module-level procedural macro attributes to instrument functions with tracing"
documentation = "https://docs.rs/crustrace"
edition = "2024"
homepage = "https://github.com/lmmx/crustrace"
license = "MIT OR Apache-2.0"
name = "crustrace"
readme = "../README.md"
repository = "https://github.com/lmmx/crustrace"
version = "0.1.0"

[lib]
proc-macro = true

[dependencies]
crustrace-core = { workspace = true }
proc-macro2.workspace = true

[dev-dependencies]
cargo-husky = { workspace = true, features = ["user-hooks"] }
insta.workspace = true
rust-format.workspace = true
tracing.workspace = true
tracing-subscriber.workspace = true
