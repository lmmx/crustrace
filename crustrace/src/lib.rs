use proc_macro::TokenStream;
use proc_macro2::TokenStream as TokenStream2;

#[proc_macro_attribute]
pub fn trace_all(_args: TokenStream, input: TokenStream) -> TokenStream {
    let input2: TokenStream2 = input.into();
    let output = crustrace_core::trace_all_impl(input2);
    output.into()
}