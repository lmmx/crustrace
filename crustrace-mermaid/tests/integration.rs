#[cfg(test)]
mod tests {
    use crustrace::instrument;
    use crustrace_mermaid::*;
    use tracing::subscriber::set_default;
    use tracing_subscriber::prelude::*;

    #[instrument]
    fn inner(x: i32, y: i32) -> i32 {
        x + y
    }

    #[instrument]
    fn outer(a: i32, b: i32) -> i32 {
        let r1 = inner(a + 1, b / 10);
        let r2 = inner(a * 2, b / 20);
        r1 + r2
    }

    #[test]
    fn snapshot_merge_by_name() {
        let layer = MermaidLayer::with_mode(GroupingMode::MergeByName).without_auto_flush();
        let subscriber = tracing_subscriber::registry().with(layer.clone());

        let _guard = set_default(subscriber); // scoped subscriber
        outer(10, 20);

        insta::assert_snapshot!(layer.render());
    }

    #[test]
    fn snapshot_unique_per_call() {
        let layer = MermaidLayer::with_mode(GroupingMode::UniquePerCall).without_auto_flush();
        let subscriber = tracing_subscriber::registry().with(layer.clone());

        let _guard = set_default(subscriber); // scoped subscriber
        outer(10, 20);

        insta::assert_snapshot!(layer.render());
    }
}
