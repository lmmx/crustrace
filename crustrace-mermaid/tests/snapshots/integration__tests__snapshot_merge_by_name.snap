---
source: crustrace-mermaid/tests/integration.rs
expression: layer.render()
---
flowchart TD
subgraph Params1[" "]
  P2_0["a = 10"]:::data
  P2_1["b = 20"]:::data
  P2_0 --- P2_1
end
F1["outer()"]:::func
Params1 --> F1
subgraph innerCalls["inner(...)"]
  direction TB
subgraph Params2[" "]
  P3_0["x = 11"]:::data
  P3_1["y = 2"]:::data
  P3_0 --- P3_1
end
F2["inner()"]:::func
Params2 --> F2
subgraph Params3[" "]
  P4_0["x = 20"]:::data
  P4_1["y = 1"]:::data
  P4_0 --- P4_1
end
F3["inner()"]:::func
Params3 --> F3
end
  F1 --> Params2
  F1 --> Params3

classDef func fill:#c6f6d5,stroke:#2f855a,stroke-width:2px,color:#22543d;
classDef data fill:#bee3f8,stroke:#2b6cb0,stroke-width:1.5px,color:#1a365d;
classDef params fill:#fefcbf,stroke:#975a16,stroke-width:1.5px,color:#744210;
class Params1,Params2,Params3 params;
